Add support for static local functions (#869)
dotnet/csharpstandard: standard/statements.md

The *identifier* of a *local_function_declaration* must be unique in its declared block scope, including any enclosing local variable declaration spaces. One consequence of this is that overloaded *local_function_declaration*s are not allowed.

> A *local_function_declaration* may include one `async` ([§15.15](classes.md#1515-async-functions)) modifier and one `unsafe` ([§23.1](unsafe-code.md#231-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a `«TaskType»` type ([§15.15.1](classes.md#15151-general)). The `unsafe` modifier uses the containing lexical scope. The `unsafe` modifier uses the containing lexical scope. The `async` modifier does not use the containing lexical scope. If the declaration includes the `static` modifier, the function is a ***static local function***; otherwise, it is a ***non-static local function***. It is a compile-time error for *type_parameter_list* or *formal_parameter_list* to contain *attributes*.

Reviewer: Doesn't need to describe `unsafe` twice.

Reviewer: The wording "uses the containing lexical scope" seems a bit vague too.  As if, when the local function has its own `unsafe` modifier, that causes the function to use something from the lexical scope.  Can this be clarified somehow… "Any `unsafe` modifier in the containing lexical scope also applies to the local function."

Reviewer 2: I agree that the "containing lexical scope" statements aren't as clear as they could be. (Not a C# 8 feature, but let's fix it anyway).  Here's my suggestion:

```suggestion
A *local_function_declaration* may include one `async` ([§15.15](classes.md#1515-async-functions)) modifier and one `unsafe` ([§23.1](unsafe-code.md#231-general)) modifier. If the declaration includes the `async` modifier then the return type shall be `void` or a `«TaskType»` type ([§15.15.1](classes.md#15151-general)). If the declaration includes the `static` modifier, the function is a ***static local function***; otherwise, it is a ***non-static local function***. It is a compile-time error for *type_parameter_list* or *formal_parameter_list* to contain *attributes*. If the local function is declared in an unsafe context (§23.2), the local function may include unsafe code, even if the local function declaration doesn't include the `unsafe` modifier.
```

> A local function is declared at block scope. A non-static local function may capture variables from the enclosing scope while a static local function may not (so it has no access to enclosing locals, parameters, or `this`). It is a compile-time error if a captured variable is read by the body of the non-static local function but is not definitely assigned before each call to the function. The compiler shall determine which variables are definitely assigned on return ([§9.4.4.33](variables.md#94433-rules-for-variables-in-local-functions)).

Reviewer 3: ```suggestion
A local function is declared at block scope. A non-static local function may capture variables from the enclosing scope while a static local function shall not (so it has no access to enclosing locals, parameters, non-static local functions, or `this`). It is a compile-time error if a captured variable is read by the body of a non-static local function but is not definitely assigned before each call to the function. The compiler shall determine which variables are definitely assigned on return ([§9.4.4.33](variables.md#94433-rules-for-variables-in-local-functions)).
```

> A local function is declared at block scope, and that function may capture variables from the enclosing scopes. A non-static local function may capture variables from the enclosing scope while a static local function may not (so it has no access to enclosing locals, parameters, or `this`). It is a compile-time error if a captured variable is read by the body of the non-static local function but is not definitely assigned before each call to the function. The compiler shall determine which variables are definitely assigned on return ([§9.4.4.33](variables.md#94433-rules-for-variables-in-local-functions)).

Reviewer 2: I'd remove the second clause from the first sentence. It's repeated in the next sentence:

```suggestion
A local function is declared at block scope. A non-static local function may capture variables from the enclosing scope while a static local function may not (so it has no access to enclosing locals, parameters, or `this`). It is a compile-time error if a captured variable is read by the body of the non-static local function but is not definitely assigned before each call to the function. The compiler shall determine which variables are definitely assigned on return ([§9.4.4.33](variables.md#94433-rules-for-variables-in-local-functions)).
```

If the type of the argument to a local function is `dynamic`, the function to be called must be resolved at compile time, not runtime.

A local function shall not be used in an expression tree.

A static local function

- May reference static members, type parameters, constant definitions and static local functions from the enclosing scope.
- Shall not reference `this` or `base` nor instance members from an implicit `this` reference, nor local variables, parameters, or non-static local functions from the enclosing scope. However, all these are permitted in a `nameof()` expression.

> Overload resolution of a call within a local function is not affected by whether the local function is static.

Reviewer 3: I think this is correct:

```csharp
class Program
{
    public void Method(int x = 10)
    {
    }

    static void Method(string y = "")
    {
    }

    void M()
    {
        StaticLocal();
        NonStaticLocal();
        static void StaticLocal()
        {
            Method();
        }

        void NonStaticLocal()
        {
            Method();
        }
    }
}
```
Both calls to Method cause errors of "The call is ambiguous between the following methods or properties: 'Program.Method(int)' and 'Program.Method(string)'"

Reviewer 3: Next attempt:

```csharp
class Program
{
    public void Method(string x)
    {
    }

    static void Method(object x)
    {
    }

    void M()
    {
        StaticLocal();
        NonStaticLocal();
        static void StaticLocal()
        {
            Method("");
        }

        void NonStaticLocal()
        {
            Method("");
        }
    }

    static void StaticM()
    {
        Method("");
    }
}
```

Here, calling `Method("")` from `StaticM` is fine, but not from `StaticLocal`.

Reviewer 3: Interestingly, if you copy `StaticLocal` into `StaticM` (and call it from `StaticM`) then it's fine.

> 
> Overload resolution of a call within a local function is not affected by whether the local function is static.

Reviewer 3: ```suggestion
```

Removing because we believe that the fact that it's currently true is a bug rather than a feature, and we may want to fix the compiler later.

If the type of the argument to a local function is `dynamic`, the function to be called must be resolved at compile time, not runtime.

A local function shall not be used in an expression tree.

A static local function

> - May reference static members, type parameters, constant definitions and static local functions from the enclosing scope.

Reviewer 4: ```suggestion
- May reference static members, type parameters, constant definitions and static local functions from the enclosing scope;
```

- Shall not reference `this` or `base` nor instance members from an implicit `this` reference, nor local variables, parameters, or non-static local functions from the enclosing scope. However, all these are permitted in a `nameof()` expression.

Overload resolution of a call within a local function is not affected by whether the local function is static.

> Removing the `static` modifier from a local function in a valid program does not change the meaning of the program.

Reviewer 3: ```suggestion
> *Note:* Removing the `static` modifier from a local function in a valid program does not change the meaning of the program. *end note*
```

> 
> Removing the `static` modifier from a local function in a valid program does not change the meaning of the program.

Reviewer 3: ```suggestion
```

If the type of the argument to a local function is `dynamic`, the function to be called must be resolved at compile time, not runtime.

> A static local function

Reviewer 2: Move expression tree comment to apply to all local functions:

```suggestion
A local function shall not be invoked in an expression tree.

A static local function
```

If the type of the argument to a local function is `dynamic`, the function to be called must be resolved at compile time, not runtime.

> A local function shall not be invoked in an expression tree.

Reviewer 3: ```suggestion
A local function shall not be used in an expression tree.
```

> - May not reference instance members from an implicit or explicit `this` or `base` reference.

Reviewer 3: ```suggestion
- Shall not reference instance members from an implicit or explicit `this` or `base` reference.
```

> - Shall not reference instance members from an implicit or explicit `this` or `base` reference.

Reviewer 3: ```suggestion
- Shall not reference instance members from an implicit or explicit `this` or `base` reference.
- Shall not reference local variables, parameters, or non-static local functions from the enclosing scope.
```

> - May reference `static` members from the enclosing scope.

Reviewer 3: ```suggestion
- May reference static members, type parameters, constant definitions and static local functions from the enclosing scope.
```

> - May reference `constant` definitions from the enclosing scope.

Reviewer 3: ```suggestion
- May reference constant definitions from the enclosing scope.
```

Reviewer 3: (Maybe we should remove the backticks from around static as well?)

> - May reference constant definitions from the enclosing scope.

Reviewer 3: ```suggestion
```

(This is now in 613.)

> - May use `nameof()` to reference locals, parameters, or `this` or `base` from the enclosing scope.

Reviewer 3: ```suggestion
- May use `nameof()` to reference locals, parameters, non-static local functions, `this` or `base` from the enclosing scope.
```

> - May capture state from an enclosing static local function, but may not capture state outside the enclosing static local function. The same rules apply to a lambda.

Reviewer 3: ```suggestion
- May capture state from an enclosing static local function, but may not capture state outside the enclosing static local function. The same rules apply to a lambda expression.
```

This final sentence feels weird though - are these rules for lambda expressions that are enumerated somewhere else? If so, let's link to them. It not, why not?

> - May capture state from an enclosing static local function, but may not capture state outside the enclosing static local function. The same rules apply to a lambda.

Reviewer 3: Earlier we state:

> A non-static local function may capture variables from the enclosing scope while a static local function may not

Doesn't that contradict "May capture state from an enclosing static local function"?

Reviewer 2: yes.

> - May capture state from an enclosing static local function, but may not capture state outside the enclosing static local function. The same rules apply to a lambda.

Reviewer 2: This is incorrect.  A `static` local function can't capture state.

This should read that a non-static local function can capture state declared in an enclosing static local function:

```csharp
static void M()
{
    OuterM();
    static void OuterM()
    {
        var msg = "hello";
        Console.WriteLine("Hello, World!");

        InnerM();
        void InnerM() // not static
        {
            Console.WriteLine(msg);
        }
    }
}
```

Reviewer 3: Great - @BillWagner are you happy to propose a change for this?

> - May capture state from an enclosing static local function, but may not capture state outside the enclosing static local function. The same rules apply to a lambda.

Reviewer 2: I think we can simply remove this line:

```suggestion
```

> - Shall not reference instance members from an implicit or explicit `this` or `base` reference.
> - Shall not reference local variables, parameters, or non-static local functions from the enclosing scope.
> - May reference static members, type parameters, constant definitions and static local functions from the enclosing scope.
> - May use `nameof()` to reference locals, parameters, non-static local functions, `this` or `base` from the enclosing scope.

Reviewer 4: ```suggestion
- May reference static members, type parameters, constant definitions and static local functions from the enclosing scope.
- Shall not reference `this` or `base` nor instance members from an implicit `this` reference, nor local variables, parameters, or non-static local functions from the enclosing scope. However, all these are permitted in a `nameof()` expression.
```

> - May not be invoked in an expression tree.

Reviewer 3: ```suggestion
```

Reviewer 3: Is it valid to invoke a non-static local function from an expressoin tree?

Reviewer 2: No, it's not. Local functions (static or not) aren't allowed in expression trees. We missed this in 7.3
